pkg/operator/image.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2020 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/blang/semver"
+)
+
+// ImageConfig contains config for creating a container image URL
+type ImageConfig struct {
+	Image            *string
+	DefaultBaseImage string
+	BaseImage        string
+	DefaultVersion   string
+	Version          string
+	Tag              string
+	SHA              string
+}
+
+// ContainerImageURL determines the appropriate container image URL based
+// on the data coming from default values, operator CLI args, and user
+// configured fields in the CRD.
+//
+// If the `Image` field is not empty, this takes priority over other
+// settings.  If the given `Image` URL does not contain a version,
+// then the configured version or the default version is appended to
+// the image URL.
+//
+// If no `Image` is specified, an image URL is created by combining the
+// appropriate `BaseImage` with the `Version`, `Tag`, and `SHA`, fields.
+//
+// Returns the image URL of the component.
+func ContainerImageURL(c *ImageConfig) string {
+	version := c.DefaultVersion
+	if c.Version != "" {
+		version = c.Version
+	}
+
+	if c.Image != nil && *c.Image != "" {
+		return ApplyImageVersion(*c.Image, version)
```

Comment:
we should make image and version handling completely distinct (as I believe it was before)

Reply:
I tried to keep the output the same as the previous version handling.  The only difference should be the fix to #3110 which changes the image URL only if the URL string doesn't already contain a version.  If `c.Image` here contains a tag or sha then it will be returned unmodified.

The version calculation (for flags and other changes) is done separately in the `ChooseVersion` function.  First we take the user specified version if it's set, then check if the image URL contains a valid version string, and lastly use the default version.

So I think hopefully this is already doing what you are asking for?

Reply:
I think what's happening in https://github.com/coreos/prometheus-operator/issues/3110 is the correct behavior. It was a mistake to ever have `version` be a deciding factor in which image is being used. We should decouple this as much as we can without breaking existing functionality. `version` should describe the way the resources are generated by the Prometheus Operator, or on some way say "which version" is behind the opaque `image` that is provided.

Reply:
Ok, I think that's reasonable and simplifies the code.  It might be a little less intuitive to the users, but we can probably handle that with docs.
Just to confirm some possible use cases:
1. User specifies an image URL such as `prom/prometheus:v1.7.2`, the `version` in terms of behaviour will be the default of `v2.16.0` unless the user also sets the `version` field.
2. User specifies only the `version` field to `v1.7.2`, the default image of `quay.io/prometheus/prometheus:v2.16.0` will still be used.

Reply:
Yes I think we should handle this with docs and we should probably remove the defaulting behavior completely (eventually, not now) and enforce version to be set by the user.